packages/core/src/Renderer.ts

```diff
@@ -222,12 +131,24 @@ export class Renderer {
      *
      * @param context
      */
-    _renderVNode(context: RenderingContext): RenderingContext {
-        if (isChar(context.currentVNode)) {
-            context = this._renderTextNode(context);
-        } else {
-            context = this._renderElement(context);
-        }
+    _renderNext(context: RenderingContext): RenderingContext {
+        const rendering = context.currentVNode.render<HTMLRendering>('html');
+        this._addRenderedNodes(
+            Array.from(rendering.fragment.childNodes) as Node[],
+            rendering.vNodes,
+            context.parentNode,
+        );
+        context.currentVNode = rendering.vNodes[rendering.vNodes.length - 1];
         return context;
     }
+    // Append the rendered nodes to the DOM and map them.
+    _addRenderedNodes(domNodes: Node[], vNodes: VNode[], parent: Node): void {
+        domNodes.forEach(renderedNode => {
+            vNodes.forEach((vNode, index) => {
```

Comment:
O(n<sup>2</sup>) * recursion. Though granted `n` in this case should remain quite small.
This is solved if `render` returns a rendering context.